### PR TITLE
Auto create time layers in geoserver_publisher

### PR DESCRIPTION
**Pull request by chrismayer**
This creates time enabled WMS layers for the given ImageMosaic stores if not already existing in the GeoServer. If existing the TIME dimension is activated for the given layer.

Also ensures that the password for the DB in the datastore.properties is injected from the corresponding secret.

### CODE REVIEW

File: geoserver_publisher/index.js [hunk 1]
```diff
@@ -121,6 +124,40 @@ async function checkIfCoverageStoresExist(rasterMetaInf) {
   }
 }
 
+/**
+ * Creates a time-enabled layer in the GeoServer for the given raster mosaic if
+ * not existing by GeoServer REST-API.
+ *
+ * @param {Object} rasterMetaInf
```
Comment (JakobMiksch) [inline geoserver_publisher/index.js:131]: A description of the input would be helpful

File: geoserver_publisher/index.js [hunk 2]
```diff
@@ -121,6 +124,40 @@ async function checkIfCoverageStoresExist(rasterMetaInf) {
...
+ */
+async function createRasterTimeLayers (rasterMetaInf) {
+  const ws = rasterMetaInf.workspace;
+  const covStore = rasterMetaInf.coverage_store;
+  const srs = 'EPSG:3035'; //TODO check if defined in DB
```
Comment (JakobMiksch) [inline geoserver_publisher/index.js:136]: whitespace

```suggestion
  const srs = 'EPSG:3035'; // TODO check if defined in DB
```

File: geoserver_publisher/index.js [hunk 3]
```diff
@@ -121,6 +124,40 @@ async function checkIfCoverageStoresExist(rasterMetaInf) {
...
+
+  if (!layer) {
+    console.info(`Creating layer "${ws}:${layerName}" in store "${covStore}"`);
+    // publishDbRaster (workspace, coverageStore, nativeName, name, title, srs, enabled)
+    const layerCreated = await grc.layers.publishDbRaster(ws, covStore, nativeName, layerName, layerTitle, srs, true);
+    verboseLogging(`Layer "${ws}:${layerName}" created successfully?`, layerCreated);
+
+    console.info(`Enabling time for layer "${ws}:${layerName}"`);
+    const timeEnabled = await grc.layers.enableTimeCoverage(ws, covStore, layerName, 'DISCRETE_INTERVAL', 3600000, 'MAXIMUM');
+    verboseLogging(`Time dimension  for layer "${ws}:${layerName}" successfully enabled?`, timeEnabled);
```
Comment (JakobMiksch) [inline geoserver_publisher/index.js:154]: superfluous whitespace
```suggestion
    verboseLogging(`Time dimension for layer "${ws}:${layerName}" successfully enabled?`, timeEnabled);
```

File: geoserver_publisher/index.js [hunk 4]
```diff
@@ -121,6 +124,40 @@ async function checkIfCoverageStoresExist(rasterMetaInf) {
...
+  const nativeName = rasterMetaInf.properties_path;
+  const layerTitle = covStore;
+
+  verboseLogging(`Checking existence for layer ${ws}:${layerName} in coverage store ${covStore} (native name: ${nativeName})`);
+
+  const layer = await grc.layers.get(covStore);
+
+  if (!layer) {
+    console.info(`Creating layer "${ws}:${layerName}" in store "${covStore}"`);
+    // publishDbRaster (workspace, coverageStore, nativeName, name, title, srs, enabled)
```
Comment (JakobMiksch) [inline geoserver_publisher/index.js:148]: maybe a quick comment that this should describe the input of the function below. 
On first sight, I thought this was a commented functionCall

**Review by JakobMiksch — APPROVED** [review]
The code looks fine. Some minor remarks.

**Review by JakobMiksch — CHANGES_REQUESTED** [review]
I need to test it locally before I can approve finally

**Review by JakobMiksch — APPROVED** [review]
I finally tested it locally on my machine. The functions seem to be called, but since i do not have rasters they also cannot be published. @chrismayer please merge at will

**Comment by chrismayer** [comment]
Thanks for the review. I'd like to wait what @jdk24 says to this.

**Review by jdk24 — CHANGES_REQUESTED** [review]
The [image mosaic template ](https://github.com/meggsimum/sauber-sdi-docker/blob/master/geoserver_publisher/gs-img-mosaic-tpl/datastore.properties#L7) which is read [here](https://github.com/meggsimum/sauber-sdi-docker/pull/61/files#diff-ca92621faca806ac3b80fe78d54e869f91ea735480602d00a3e730e33cddc7ebR92) still contains the early-days standard DB password. 

On attempting to create a new image mosaic, this results in `failed to create reader from file` from GeoServer so far. 

Geoserver_publisher probably needs to have access to the relevant secret and insert it into the template.

**Comment by JakobMiksch** [comment]
@jdk24 @chrismayer I will approach this issue of adding the secret

**Comment by chrismayer** [comment]
> Geoserver_publisher probably needs to have access to the relevant secret and insert it into the template.

Implemented and rebased the original commit. Please have another look.

**Comment by chrismayer** [comment]
> https://github.com/meggsimum/sauber-sdi-docker/blob/b95624dabe80be55b120733f42741f979b69b6c2/docker-stack.yml#L302-L305
> 
> Please add
> 
> ```
>  - sauber_manager_password
> ```

Nice catch @jdk24 :gun: I addressed this and also the remarks of @JakobMiksch. 

> ## **Edit:**
> Updating `docker-stack.yml` does allow the raster publisher to create an ImgMosaic store & publish a time-enabled raster layer from incoming raster data. The bbox and time dimension are set correctly.
> 
> On trying to query the layer however, I get [this error](https://github.com/meggsimum/sauber-sdi-docker/files/6732112/wms-6.txt). This also happens when
> 
> 1. manually creating a layer in the automatically created ImgMosaic store
> 2. manually creating a ImgMosaic store & layer
> 3. manually publishing the incoming raster as simple GTIFF layer
> 
> all after cleaning geoserver_mnt, deleting stack volumes, restarting, etc. The files exist on the server and the PostGIS index looks sound.

So since this seems to be unrelated to this PR because it also appears when manually creating an ImageMosaic store & layer (as described in 2.) I vote for merging this, if you are OK. The error described above should be documented in a separate Issue and be handled in a separate PR.

**Review by jdk24 — APPROVED** [review]
LGTM. ImgStore created, Layer published and time dimension enabled. Please merge at will!

**Comment by chrismayer** [comment]
Thanks for your reviews! I'll merge now.

**Pull request merged**